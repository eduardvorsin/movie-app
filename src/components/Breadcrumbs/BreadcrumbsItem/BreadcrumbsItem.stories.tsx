import type { Meta, StoryObj } from '@storybook/react';
import BreadcrumbsItem from './BreadcrumbsItem';

const meta: Meta<typeof BreadcrumbsItem> = {
	title: 'UI/Breadcrumbs/BreadcrumbsItem',
	tags: ['autodocs'],
	argTypes: {
		children: {
			description: 'The text content of the element',
		},
		className: {
			description: 'Additional classes for the component'
		},
		onClick: {
			control: false,
			description: 'The callback that is called when the component is clicked',
		},
		href: {
			description: 'The link to the page to which the transition will be made when clicking on the component',
		},
		testId: {
			description: 'Id for testing the component',
		},
		isLastItem: {
			description: 'Determines whether the link is the last in the list',
		},
		truncationWidth: {
			description: 'Sets the maximum width of the component at which the text will be cropped',
		},
		target: {
			control: false,
			description: 'How to display a page when clicking on a link',
		},
	},
	component: BreadcrumbsItem,
	parameters: {
		docs: {
			description: {
				component: 'The BreadcrumbsItem component is a part of the breadcrumbs system and is used to represent a single element in this navigation chain.',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof BreadcrumbsItem>;

export const Default: Story = {
	args: {
		children: 'item',
		href: '/',
	},
	parameters: {
		docs: {
			description: {
				story: 'The default state of the component',
			},
		},
	},
};

export const TrunctionWidth: Story = {
	args: {
		children: 'A very long element',
		href: '/',
		truncationWidth: 100,
	},
	parameters: {
		docs: {
			description: {
				story: 'The value in pixels after which the text in the component will be cropped',
			},
		},
	},
};

export const isLastItem: Story = {
	args: {
		children: 'Last item',
		href: '/',
		isLastItem: true,
	},
	parameters: {
		docs: {
			description: {
				story: 'Since the last link in the list points to the current page, we use this prop to make it inactive for users',
			},
		},
	},
};

export const EventCallbacks: Story = {
	argTypes: {
		onClick: {
			action: 'Clicked',
		},
	},
	args: {
		children: 'item',
		href: '/',
	},
	parameters: {
		docs: {
			description: {
				story: 'A callback function can be provided to the onClick event.',
			},
		},
	},
};