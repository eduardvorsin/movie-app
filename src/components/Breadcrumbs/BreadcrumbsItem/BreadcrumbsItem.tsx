'use client';
import { AnchorHTMLAttributes, MouseEventHandler, ReactNode } from 'react';
import Link from '@/components/Link/Link';
import { GeneralProps } from '@/types/shared';

type Props = {
	children: ReactNode,
	href?: string,
	onClick?: MouseEventHandler<HTMLAnchorElement>,
	truncationWidth?: number,
	target?: AnchorHTMLAttributes<HTMLButtonElement>['target'],
} & GeneralProps;

export default function BreadcrumbsItem({
	children,
	className,
	href = '',
	onClick,
	truncationWidth,
	target,
	testId,
	...props
}: Props) {

	return (
		<Link
			style={truncationWidth ? { 'maxWidth': `${truncationWidth}px` } : undefined}
			className={`inline-flex text-200 ${className}`}
			href={href}
			onClick={onClick}
			target={target}
			testId={testId}
			{...props}
		>
			<span className={truncationWidth ? 'truncate' : ''}>
				{children}
			</span>
		</Link>
	);
};