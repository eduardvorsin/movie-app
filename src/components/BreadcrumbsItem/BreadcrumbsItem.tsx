'use client';
import { AnchorHTMLAttributes, MouseEventHandler, ReactNode } from 'react';
import Link from '../Link/Link';

type Props = {
	text: string,
	className?: string,
	href?: string,
	iconBefore?: ReactNode,
	iconAfter?: ReactNode,
	onClick?: MouseEventHandler<HTMLAnchorElement>,
	truncationWidth?: number,
	target?: AnchorHTMLAttributes<HTMLButtonElement>['target'],
	testId?: string,
};

export default function BreadcrumbsItem({
	text,
	className,
	href = '',
	iconBefore,
	iconAfter,
	onClick,
	truncationWidth,
	target,
	testId,
}: Props) {

	const classes = [
		truncationWidth ? 'truncate' : '',
		className,
	].join(' ');

	const Children: JSX.Element =
		(
			<>
				{iconBefore && (
					<span className='w-6 h-6 mr-1'>
						{iconBefore}
					</span>
				)}

				<span>
					{text}
				</span>

				{iconAfter && (
					<span className='w-6 h-6 ml-1'>
						{iconAfter}
					</span>
				)}
			</>
		);

	return (
		<Link
			style={truncationWidth ? { 'maxWidth': `${truncationWidth}px` } : undefined}
			className={classes}
			href={href}
			onClick={onClick}
			target={target}
			data-testid={testId}
		>
			{Children}
		</Link>
	);
};

