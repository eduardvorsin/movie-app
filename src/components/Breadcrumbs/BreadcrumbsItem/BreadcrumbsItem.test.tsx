import { render, screen } from "@testing-library/react";
import BreadcrumbsItem from "./BreadcrumbsItem";
import userEvent from "@testing-library/user-event";

describe('BreadcrumbsItem tests', () => {
	it('is rendered correctly', () => {
		render(
			<BreadcrumbsItem
				testId='test-breadcrumbs-item'
			>
				movies
			</BreadcrumbsItem>
		);

		expect(screen.getByTestId<HTMLAnchorElement>('test-breadcrumbs-item')).toBeInTheDocument();
	});

	it('when clicking on a component, the mock function is called', async () => {
		const user = userEvent.setup();
		const mockFn = jest.fn();
		render(
			<BreadcrumbsItem
				onClick={mockFn}
				testId='test-breadcrumbs-item'
			>
				movies
			</BreadcrumbsItem>
		);

		await user.click(screen.getByRole<HTMLAnchorElement>('link'));

		expect(mockFn).toHaveBeenCalledTimes(1);
	});

	it('is a basic snapshot', () => {
		render(
			<BreadcrumbsItem
				testId='test-breadcrumbs-item'
			>
				movies
			</BreadcrumbsItem>
		);

		expect(screen.getByTestId<HTMLAnchorElement>('test-breadcrumbs-item')).toMatchSnapshot();
	});

	it('is a snapshot with a truncationWidth of 100', () => {
		render(
			<BreadcrumbsItem
				truncationWidth={100}
				testId='test-breadcrumbs-item'
			>
				movies
			</BreadcrumbsItem>
		);

		expect(screen.getByTestId<HTMLAnchorElement>('test-breadcrumbs-item')).toMatchSnapshot();
	});
});