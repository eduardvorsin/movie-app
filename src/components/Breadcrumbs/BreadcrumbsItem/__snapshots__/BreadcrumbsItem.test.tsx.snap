// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadcrumbsItem tests is a basic snapshot 1`] = `
<a
  aria-disabled="true"
  class="no-underline  hover:underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200  inline-flex text-200 undefined"
  data-testid="test-breadcrumbs-item"
  href=""
  role="link"
>
  <span
    class=""
  >
    movies
  </span>
</a>
`;

exports[`BreadcrumbsItem tests is a snapshot with a truncationWidth of 100 1`] = `
<a
  aria-disabled="true"
  class="no-underline  hover:underline transition-colors duration-150 text-blue-800 hover:text-blue-900 active:text-blue-1000 dark:text-blue-400 dark:hover:text-blue-300 dark:active:text-blue-200  inline-flex text-200 undefined"
  data-testid="test-breadcrumbs-item"
  href=""
  role="link"
  style="max-width: 100px;"
>
  <span
    class="truncate"
  >
    movies
  </span>
</a>
`;
